File: jiuzhang/surrounded-regions/surrounded-regions.cpp
/**
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
 */

class Solution {
public:
    /**
     * @param board a 2D board containing "X" and "O"
     * @return void
     */
    void surroundedRegions(vector>& board) {
        int m=board.size();
		if (m==0)
			return;
		int n=board[0].size();
		int i,j;
		for(i=0;i<m;i++)
		{
			if (board[i][0]=="O")
				process(i,0,board);
			if (board[i][n-1]=="O")
				process(i,n-1,board);
		}
		for(j=0;j<n;j++)
		{
			if (board[0][j]=="O")
				process(0,j,board);
			if (board[m-1][j]=="O")
				process(m-1,j,board);
		}
		for(i=0;i<m;i++)
			for(j=0;j<n;j++)
			{
				if(board[i][j]=="O")
					board[i][j]="X";
				else if (board[i][j]=="E")
					board[i][j]="O";
			}
    }
    void process(int i,int j,vector >& board)
	{
		int m=board.size();
		int n=board[0].size();

		typedef pair point;
		queue Q;
		Q.push(point(i,j));
		board[i][j]="E";
		while(!Q.empty())
		{
			point tmp=Q.front();
			Q.pop();
			
			int x=tmp.first,y=tmp.second;

			//extending
			if (x!=0&&board[x-1][y]=="O")
			{
				Q.push(point(x-1,y));
				board[x-1][y]="E";   //extended;
			}
			if (x!=m-1&&board[x+1][y]=="O")
			{
				Q.push(point(x+1,y));
				board[x+1][y]="E";   //extended;
			}
			if (y!=0&&board[x][y-1]=="O")
			{
				Q.push(point(x,y-1));
				board[x][y-1]="E";   //extended;
			}
			if (y!=n-1&&board[x][y+1]=="O")
			{
				Q.push(point(x,y+1));
				board[x][y+1]="E";   //extended;
			}
		}
    }
};

File: jiuzhang/surrounded-regions/surrounded-regions.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""

class Solution:
    # @param {list[list[str]]} board a 2D board containing "X" and "O"
    # @return nothing 
    def surroundedRegions(self, board):
        def fill(x, y):
            if x  m-1 or y < 0 or y > n-1 or board[x][y] != "O":
                return
            queue.append((x,y))
            board[x][y] = "D"

        def bfs(x, y):
            if board[x][y]=="O":
                queue.append((x,y))
                fill(x,y)

            while queue:
                curr = queue.pop(0)
                i, j = curr[0], curr[1]
                fill(i+1, j)
                fill(i-1, j)
                fill(i, j+1)
                fill(i, j-1)

        if len(board) == 0:
            return
        m, n , queue = len(board), len(board[0]), []
        for i in range(n):
            bfs(0, i)
            bfs(m-1, i)

        for j in range(1, m-1):
            bfs(j, 0)
            bfs(j, n-1)

        for i in range(m):
            for j in range(n):
                if board[i][j] == "D":
                    board[i][j] = "O"
                elif board[i][j] == "O":
                    board[i][j] = "X"

